File: spring/src/main/java/cardsproject/domain/marketplace/TradeDispute.java
package cardsproject.domain.marketplace;

import jakarta.persistence.*;
import java.time.LocalDateTime;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

@Entity
@Table(name = "trade_disputes")
public class TradeDispute {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Enumerated(EnumType.STRING)
    private TradeDisputeStatusType status;
    @Enumerated(EnumType.STRING)
    private TradeDisputeReasonType reason;
    private String description = "";
    private String resolution;
    private LocalDateTime openedAt;
    private LocalDateTime resolvedAt;

    // @OneToOne -> TradeTransaction, onDelete=CASCADE, relatedName=dispute
    @Column(name = "transaction_id")
    private Long transactionId;
    // @ManyToOne -> Player, onDelete=PROTECT, relatedName=disputes_opened, via=players
    @Column(name = "opened_by_id")
    private Long openedById;
    // @ManyToOne -> Player, onDelete=SET_NULL, relatedName=disputes_resolved, via=players
    @Column(name = "resolved_by_id")
    private Long resolvedById;

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public TradeDisputeStatusType getStatus() { return status; }
    public void setStatus(TradeDisputeStatusType status) { this.status = status; }
    public TradeDisputeReasonType getReason() { return reason; }
    public void setReason(TradeDisputeReasonType reason) { this.reason = reason; }
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    public String getResolution() { return resolution; }
    public void setResolution(String resolution) { this.resolution = resolution; }
    @JsonProperty("openedAt")
    public LocalDateTime getOpenedAt() { return openedAt; }
    public void setOpenedAt(LocalDateTime openedAt) { this.openedAt = openedAt; }
    @JsonProperty("resolvedAt")
    public LocalDateTime getResolvedAt() { return resolvedAt; }
    public void setResolvedAt(LocalDateTime resolvedAt) { this.resolvedAt = resolvedAt; }
    public Long getTransactionId() { return transactionId; }
    public void setTransactionId(Long transactionId) { this.transactionId = transactionId; }
    public Long getOpenedById() { return openedById; }
    public void setOpenedById(Long openedById) { this.openedById = openedById; }
    public Long getResolvedById() { return resolvedById; }
    public void setResolvedById(Long resolvedById) { this.resolvedById = resolvedById; }

    // ── Business operations ──────────────────────────────────────────
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void escalate() {
        // TODO: implement escalate
    }
    public void resolve(String resolutionText) {
        // TODO: implement resolve
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void closeResolved() {
        // TODO: implement closeResolved
    }
    @com.fasterxml.jackson.annotation.JsonIgnore
    public void review() {
        // TODO: implement review
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static final java.util.Map<TradeDisputeStatusType, java.util.List<TradeDisputeStatusType>> ALLOWED_TRANSITIONS =
        java.util.Map.ofEntries(
        java.util.Map.entry(TradeDisputeStatusType.OPEN, java.util.List.of(TradeDisputeStatusType.UNDERREVIEW)),
        java.util.Map.entry(TradeDisputeStatusType.UNDERREVIEW, java.util.List.of(TradeDisputeStatusType.RESOLVED, TradeDisputeStatusType.ESCALATED)),
        java.util.Map.entry(TradeDisputeStatusType.ESCALATED, java.util.List.of(TradeDisputeStatusType.RESOLVED))
        );

    public void assertTransition(TradeDisputeStatusType to) {
        java.util.List<TradeDisputeStatusType> allowed = ALLOWED_TRANSITIONS.getOrDefault(this.getStatus(), java.util.List.of());
        if (!allowed.contains(to)) {
            throw new IllegalStateException("Transition " + this.getStatus() + " -> " + to + " not allowed");
        }
    }
}

File: phoenix/lib/cards_project/marketplace/trade_dispute.ex
defmodule CardsProject.Marketplace.TradeDispute do
  use Ecto.Schema
  import Ecto.Changeset

  schema "trade_disputes" do
    field :status, :string
    field :reason, :string
    field :description, :string
    field :resolution, :string
    field :opened_at, :naive_datetime
    field :resolved_at, :naive_datetime
    belongs_to :transaction, CardsProject.Marketplace.TradeTransaction
    belongs_to :opened_by, CardsProject.Players.Player
    belongs_to :resolved_by, CardsProject.Players.Player

    timestamps()
  end

  @doc false
  def changeset(record, attrs) do
    record
    |> cast(attrs, [:description, :opened_at, :status, :reason, :resolution, :resolved_at, :transaction_id, :opened_by_id, :resolved_by_id])
    |> validate_required([:description, :opened_at])
    |> validate_inclusion(:status, ["Open", "UnderReview", "Resolved", "Escalated"])
    |> validate_inclusion(:reason, ["ItemNotReceived", "ItemNotAsDescribed", "FraudSuspected", "Other"])
  end

  # ── Business operations ────────────────────────────────────────────

  def escalate(_record) do
    # TODO: implement TradeDispute.escalate
    :ok
  end

  def resolve(_record, _resolution_text) do
    # TODO: implement TradeDispute.resolve
    :ok
  end

  def close_resolved(_record) do
    # TODO: implement TradeDispute.close_resolved
    :ok
  end

  def review(_record) do
    # TODO: implement TradeDispute.review
    :ok
  end

  # ── Lifecycle state machine ─────────────────────────────────────────
  @allowed_transitions %{
    "Open" => ["UnderReview"],
    "UnderReview" => ["Resolved", "Escalated"],
    "Escalated" => ["Resolved"]
  }

  def assert_transition(%__MODULE__{status: current}, to) do
    allowed = Map.get(@allowed_transitions, current, [])
    if to in allowed do
      :ok
    else
      {:error, "Transition #{current} -> #{to} not allowed"}
    end
  end
end
